lib/improver/tests/spotdata/spotdata/test_site_data.py

```diff
@@ -0,0 +1,300 @@
+        """
+        Substitute a dictionary key in a list of dictionaries.
+
+        """
+        for dictionary in list_of_dictionaries:
+            dictionary[newkey] = dictionary.pop(oldkey)
+
+    def save_json(self, data):
+        """
+        Save the passed in data as a json file.
+
+        """
+        self.site_path = self.data_directory + '/site_data.json'
+        ff = open(self.site_path, 'w')
+        json.dump(data, ff, sort_keys=True, indent=4,
+                  separators=(',', ': ',))
+        ff.close()
+
+
+class Test_from_file(Test_ImportSiteData):
+    """Test function used for loading site data from json files."""
+
+    method = 'from_file'
+
+    def test_complete(self):
+        """
+        Test loading a valid json file containing site details.
+        File formatted to represent current BestData input.
+
+        """
+        expected = self.site_data
+        self.save_json(self.site_data)
+
+        self.change_key(expected, 'gmtoffset', 'utc_offset')
+        result = Plugin(self.method).process(self.site_path)
+
+        for key in self.variables:
+            expected_vals = np.array([site[key] for site in expected])
+            result_vals = np.array([result[index][key] for index in result])
+            self.assertArrayEqual(expected_vals, result_vals)
+
+        # bestdata_id used as key in resulting dictionary.
+        self.assertArrayEqual(
+            np.array([site['bestdata_id'] for site in expected]),
+            np.array([index for index in result]))
+
+    def test_no_latitude(self):
+        """
+        Test loading a valid json file containing site details,
+        but missing compulsary latitude data.
+
+        """
+        for site in self.site_data:
+            site.pop('latitude')
+        self.save_json(self.site_data)
+
+        msg = 'longitude and latitude must be defined'
+        with self.assertRaisesRegexp(KeyError, msg):
+            Plugin(self.method).process(self.site_path)
+
+    def test_no_longitude(self):
+        """
+        Test loading a valid json file containing site details,
+        but missing compulsary longitude data.
+
+        """
+        for site in self.site_data:
+            site.pop('longitude')
+        self.save_json(self.site_data)
+
+        msg = 'longitude and latitude must be defined'
+        with self.assertRaisesRegexp(KeyError, msg):
+            Plugin(self.method).process(self.site_path)
+
+    def test_unequal_lat_lon(self):
+        """
+        Test loading a valid json file containing site details,
+        but missing compulsary longitude data for some sites.
+
+        """
+        self.site_data[0].pop('longitude')
+        self.save_json(self.site_data)
+        expected = 4  # No. of sites expected.
+        result = Plugin(self.method).process(self.site_path)
+
+        self.assertEqual(len(result), expected)
+        self.assertNotIn(16, np.array([index for index in result]))
+
```

Comment:
Do we get any messages if a site doesn't  have a latitude or a longitude or does it just get silently thrown out. Some sort of Warning would be nice; if there are a lot of failures one super message rather a message for each failing site would be good.

self.site_data[0]['bestdata_id'] would be better than the mysterious number 16